### DISCUSSION TITLE
Settings not saving

### DISCUSSION THREAD

**Discussion by shoot991** [cgiesche/streamdeck-homeassistant#323] [Help and Support] [answered]
Unable to save settings. New install of stream deck and plugin. Tried reinstalling. SD: 6.8.1 (21263) Plugin: 3.5.2

ws://192.168.0.228:8123/api/websocket
eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiIxY2M2OGYxMjgxMDc0YTNiODNiMjZjODY2ZmQ5MmYxMCIsImlhdCI6MTczNTMzMzA3NywiZXhwIjoyMDUwNjkzMDc3fQ.rpBjLft73WRb8m7PYOegHzzRBBwgNL9wb_mAFpVfu2w

```Updating manifest.
pi-d6f92943.js:3 Failed to download updated manifest.yml: TypeError: Cannot set properties of undefined (setting 'manifest')
js-yaml-ad175854.js:1 Parsing version 1 settings: {"version":1}
js-yaml-ad175854.js:1 Parsing version 2 settings: {"version":2,"display":{"buttonTitle":"{{friendly_name}}","enableServiceIndicator":true},"button":{"service":{"domain":"","name":"","data":""},"serviceLongPress":{"domain":"","name":"","data":""}}}
js-yaml-ad175854.js:1 Parsing version 3 settings: {"version":3,"display":{"buttonTitle":"{{friendly_name}}","enableServiceIndicator":true},"button":{"serviceLongPress":{"serviceId":"","entityId":"","serviceData":""},"serviceShortPress":{"serviceId":"","entityId":"","serviceData":""}}}
js-yaml-ad175854.js:1 Parsing version 4 settings: {"version":4,"display":{"buttonTitle":"{{friendly_name}}","enableServiceIndicator":true},"button":{"serviceLongPress":{"serviceId":"","entityId":"","serviceData":""},"serviceShortPress":{"serviceId":"","entityId":"","serviceData":""},"serviceRotation":{"serviceId":"","entityId":"","serviceData":""},"serviceTap":{"serviceId":"","entityId":"","serviceData":""}},"controllerType":"Keypad","rotationTickMultiplier":1,"rotationTickBucketSizeMs":300}
js-yaml-ad175854.js:1 Parsing version 5 settings: {"version":5,"display":{"buttonTitle":"{{friendly_name}}","enableServiceIndicator":true,"iconSettings":"PREFER_PLUGIN"},"button":{"serviceLongPress":{"serviceId":"","entityId":"","serviceData":""},"serviceShortPress":{"serviceId":"","entityId":"","serviceData":""},"serviceRotation":{"serviceId":"","entityId":"","serviceData":""},"serviceTap":{"serviceId":"","entityId":"","serviceData":""}},"controllerType":"Keypad","rotationTickMultiplier":1,"rotationTickBucketSizeMs":300}
js-yaml-ad175854.js:1 Sending HomeAssistant command:
 {
  "id": 2,
  "type": "get_states"
}
js-yaml-ad175854.js:1 Sending HomeAssistant command:
 {
  "id": 3,
  "type": "get_services"
}
js-yaml-ad175854.js:1 Sending HomeAssistant command:
 {
  "id": 2,
  "type": "get_states"
}
js-yaml-ad175854.js:1 Sending HomeAssistant command:
 {
  "id": 3,
  "type": "get_services"
}

```
Unhandled Event: deviceDidConnect
plugin-3040be3b.js:90 Got global settings.
plugin-3040be3b.js:90 Connecting to Home Assistant
js-yaml-ad175854.js:1 Parsing version 1 settings: {"version":1}
js-yaml-ad175854.js:1 Parsing version 2 settings: {"version":2,"display":{"buttonTitle":"{{friendly_name}}","enableServiceIndicator":true},"button":{"service":{"domain":"","name":"","data":""},"serviceLongPress":{"domain":"","name":"","data":""}}}
js-yaml-ad175854.js:1 Parsing version 3 settings: {"version":3,"display":{"buttonTitle":"{{friendly_name}}","enableServiceIndicator":true},"button":{"serviceLongPress":{"serviceId":"","entityId":"","serviceData":""},"serviceShortPress":{"serviceId":"","entityId":"","serviceData":""}}}
js-yaml-ad175854.js:1 Parsing version 4 settings: {"version":4,"display":{"buttonTitle":"{{friendly_name}}","enableServiceIndicator":true},"button":{"serviceLongPress":{"serviceId":"","entityId":"","serviceData":""},"serviceShortPress":{"serviceId":"","entityId":"","serviceData":""},"serviceRotation":{"serviceId":"","entityId":"","serviceData":""},"serviceTap":{"serviceId":"","entityId":"","serviceData":""}},"controllerType":"Keypad","rotationTickMultiplier":1,"rotationTickBucketSizeMs":300}
js-yaml-ad175854.js:1 Parsing version 5 settings: {"version":5,"display":{"buttonTitle":"{{friendly_name}}","enableServiceIndicator":true,"iconSettings":"PREFER_PLUGIN"},"button":{"serviceLongPress":{"serviceId":"","entityId":"","serviceData":""},"serviceShortPress":{"serviceId":"","entityId":"","serviceData":""},"serviceRotation":{"serviceId":"","entityId":"","serviceData":""},"serviceTap":{"serviceId":"","entityId":"","serviceData":""}},"controllerType":"Keypad","rotationTickMultiplier":1,"rotationTickBucketSizeMs":300}
js-yaml-ad175854.js:1 Unhandled Event: titleParametersDidChange
js-yaml-ad175854.js:1 Unhandled Event: propertyInspectorDidAppear
plugin-3040be3b.js:90 Got global settings.
plugin-3040be3b.js:90 Connecting to Home Assistant
plugin-3040be3b.js:90 Got global settings.
plugin-3040be3b.js:90 Connecting to Home Assistant
js-yaml-ad175854.js:1 Unhandled Event: propertyInspectorDidDisappear
js-yaml-ad175854.js:1 Unhandled Event: propertyInspectorDidAppear
js-yaml-ad175854.js:1 Unhandled Event: propertyInspectorDidDisappear
js-yaml-ad175854.js:1 Unhandled Event: propertyInspectorDidAppear
plugin-3040be3b.js:90 Got global settings.

**Comment by shoot991** [comment, marked as answer]
After 3 days of trying to hack at it. I figured it out. I run Windows 11 24H2 Insiders. The issue is with some network congestion algorithim in windows. 

Found this post: https://www.reddit.com/r/elgato/comments/1ftv4z1/time_plugin_no_longer_works_on_24h2/

netsh int tcp set supplemental Internet CongestionProvider=ctcp 

I did a lot to try to fix it so hopefully that is the thing that did it.

**Reply by shoot991** [reply]
You also need to restart your computer afterward

**Reply by cgiesche** [reply]
Thank you very much for identifying this issue!
It had been reported occasionally, but I wasn't able to determine the cause.
Saving the settings relies on a Stream Deck API function, so there wasn't much I could do besides passing the settings to be saved.

**Reply by shoot991** [reply]
Hopefully elgato will fix this and we can use the new algorithm. I assume the "first time users" you had mentioned in your other posts were also users like me who just got their stream deck and this was the first plugin they tried to use.

**Reply by tlusk** [reply]
I'm still on 23H2 and I'm also having issues with this plugin not saving settings.  My CongestionProvider defaulted to `cubic` and not the new `bbr2` settings from 24H2.

I tried changing it to `ctcp` to see if it would make any difference and it does not, it still won't save.

I dug into the plugin and added some debug logging and what what I can tell the websocket communication is working fine, it's calling `setGlobalSettings` with a payload containing the desired settings, but immediately after the `didReceiveGlobalSettings` event does not contain the settings at all.

```
02:06:04.338 Saving Global Settings: {"event":"setGlobalSettings","context":"85C401E7F89A0627B44C2860B21577A5","payload":{"serverUrl":"wss://<my-server-url>","accessToken":"<my-access-token>"}}
02:06:04.340 Incoming Event: {"event":"didReceiveGlobalSettings","payload":{"settings":{}}}
```

Interestingly this issue only seems to be affecting the global settings, I can see `setSettings` being called with a payload and `didReceiveSettings` actually returning the persisted payload.

For the time being I've just forked the plugin locally and hard-coded my global settings in the source to get past this.

**Reply by tlusk** [reply]
Ok I finally tracked down the issue.

The StreamDeck software handles global settings differently from the local settings because in a lot of cases they probably contain credentials like API keys or OAuth session tokens. Therefore instead of saving those settings in the config files under the `AppData/Roaming/Elgato/StreamDeck` folder, it encrypts them and stores them in the registry.

In order to do the encryption it generates an encryption key (per plugin) and stores it in the Windows Credential Manager.

The problem was (at least my problem) that the Windows Credential Manager can... run out of space (seriously!?) and not let you add new entries.  You can confirm if this is the issue by going to the Credential Manager (In the Control Panel), going to Windows Credentials, and trying to add a new "Generic Credential". In my case this was the error I got when I manually tried to add a credential:

<img src="https://github.com/user-attachments/assets/e7a1a37d-034e-48ab-afc5-6911be7d7f97" width="300">

Because of that it was unable to create a new credential to encrypt the data, and therefore wouldn't store the data. Another symptom of this issue are these two log lines in the StreamDeck0.log file when you click the "Save and connect" or "Save and reconnect" buttons in this plugin:

```
2024-12-30T13:32:04.808Z                     StreamDeck                        EGQTCredentialStore::SavePasswordWithServerAndUserName(): CredWrite() - GetLastError() 8
2024-12-30T13:32:04.808Z                     StreamDeck                        EGQTCredentialStore::GetPasswordWithServerAndUserName(): CredRead() - GetLastError() 1168
```

I was able to resolve this by following the advice of that error popup. I removed a handful of credentials in there (I can't tell you what's safe to remove and what's not, that's going to be very user specific) and then it had space to add credentials for the encryption keys.

This issue is NOT specific to the `streamdeck-homeassistant` plugin, I made a hello world plugin using the new SDK API and it also had the same issues with global credentials. I also noticed that the Elgato Discord plugin would require me to re-authenticate every single time the application was restarted, which again was caused by it not being able to save its global credentials.

I do believe Elgato should make this error should not be entirely swallowed by the Stream Deck software and not shown to the user, or have an alternative method for storing the keys as a fallback, because there really is no user feedback that something isn't working correctly. I have reported all of this to them as well through a support ticket but who knows if they'll actually make any changes.

Hope this helps anyone else with this issue.
